### PR TITLE
Refactor Resources struct and its helper functions

### PR DESCRIPTION
**Pull request by tylerslaton**
Donnie brought up a good point about why these resources aren't represented in the runtime `Resources` struct. I am adding those and also making some UX improvements for them.
### Checklist
- [X] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [X] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/acorn/pull/1199)
- [X] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [X] Commits follow [contributing guidance](https://github.com/acorn-io/acorn/blob/main/CONTRIBUTING.md#commits)
- [X] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [X] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

### CODE REVIEW

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 1]
```diff
@@ -60,87 +65,85 @@ func (current *Resources) Remove(incoming Resources, all bool) {
...
+	checkResource("Containers", current.Containers, incoming.Containers)
+	checkResource("Jobs", current.Jobs, incoming.Jobs)
+	checkResource("Volumes", current.Volumes, incoming.Volumes)
+	checkResource("Secrets", current.Secrets, incoming.Secrets)
+	checkResource("Images", current.Images, incoming.Images)
+	checkResource("Projects", current.Projects, incoming.Projects)
+
+	checkQuantityResource("VolumeStorage", current.VolumeStorage, incoming.VolumeStorage)
+	checkQuantityResource("Memory", current.Memory, incoming.Memory)
+	checkQuantityResource("Cpu", current.CPU, incoming.CPU)
```
Comment (njhale) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:93]: Mentioned in slack -- here's an alternative (take it or leave it):

```go
func (current *Resources) Fits(incoming Resources) error {
    // ...
	for _, cmp := range []struct{
	    resource string
	    limit, incoming any
	}{
	  {"Apps", current.Apps, incoming.Apps},
	  // ...
	  {"Cpu", current.CPU, incoming.CPU},
	}{
	   if gt(cmp.incoming, cmp.limit) {
	       exceeded = append(exceeded, cmp.resource)
	   }
	}
	//...
}

// gt returns true IFF a is greater than b.
func [T any]gt(a, b T) bool {
   switch v := a.(type) {
   case *resource.Quantity:
     return v.Cmp(b.(resource.Quantity)) > 0
   case int:
     return v > b.(int)
   }

   return false
}
```

Comment (njhale) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:93]: you could also do something like `runtime.Scheme`, where -- using reflection -- you register comparator functions for different types, then you wouldn't need a type switch

anyway, this is a nice exercise, but I don't feel that any of these approaches are _more_ understandable or practical than the approach being replaced by this PR.

**Review by njhale — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 2]
```diff
@@ -60,87 +65,85 @@ func (current *Resources) Remove(incoming Resources, all bool) {
...
-	if current.Volumes > 0 {
-		resources = append(resources, fmt.Sprintf("Volumes:%v", current.Volumes))
-	}
-	if current.Secrets > 0 {
-		resources = append(resources, fmt.Sprintf("Secrets:%v", current.Secrets))
-	}
-	if current.Images > 0 {
-		resources = append(resources, fmt.Sprintf("Images:%v", current.Images))
-	}
+func (current *Resources) NonEmptyString() string {
```
Comment (g-linville) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:104]: nit: function comment needs to be updated

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 3]
```diff
@@ -60,87 +65,85 @@ func (current *Resources) Remove(incoming Resources, all bool) {
...
+		if value > 0 {
+			resources = append(resources, resource)
+		}
 	}
-	if !current.CPU.IsZero() {
-		resources = append(resources, fmt.Sprintf("CPU:%v", current.CPU))
+
+	// Define function for checking quantity resources to keep code DRY
+	checkQuantityResource := func(resource string, currentVal resource.Quantity) {
+		if currentVal.IsZero() {
```
Comment (g-linville) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:116]: I think this is supposed to be inverted:

```suggestion
		if !currentVal.IsZero() {
```

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 4]
```diff
@@ -31,25 +34,27 @@ func (current *Resources) Add(incoming Resources) {
...
+	current.Projects += incoming.Projects
 
 	current.VolumeStorage.Add(incoming.VolumeStorage)
 	current.Memory.Add(incoming.Memory)
 	current.CPU.Add(incoming.CPU)
 }
 
 // Remove will remove the resources of another Resources struct from the current one.
-func (current *Resources) Remove(incoming Resources, all bool) {
+func (current *Resources) Remove(incoming Resources, removeAll bool) {
```
[the remark this answers is not in the document]
Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:45]: Fair enough, probably just me over adjusting.

File: pkg/apis/internal.admin.acorn.io/v1/resources.go (hunk 3 above)
Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go]: Yep!

File: pkg/apis/internal.admin.acorn.io/v1/resources.go (hunk 1 above)
Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:102]: This is super interesting. I do think that what you said is right in that it is cool but isn't moving the needle of understanding. If you're okay with, let's get the PR in as is?

**Pull request merged**